### DISCUSSION TITLE
No constant parameters

### DISCUSSION THREAD

**Discussion by jouslin** [pybamm-team/PyBaMM#4090] [Q&A] [answered]
I'm a French student and I need to simulate a battery cell in Python using PyBaMM. 
I have a question: some of my parameters are not constant throughout the cell and depend on the position. To address this, I use a function that depends on the position and assignes it to a parameter. However, it doesn't work because I can't extract the position and use it in the function. Indeed, I wrote this function 

def heat_transfer_coefficient(y):                                        #Where y is the position 
    if y < height  :
        return 472.487  
    elif y >= height :
        return 517.418 

parameters.update({"Total heat transfer coefficient [W.m-2.K-1]": heat_transfer_coefficient})

And I have this answers from python : 

  File "c:\Users\AJOUSLI\env\simulation.py", line 36, in heat_transfer_coefficient
    if y < height :
       ^^^^^^^
  File "c:\Users\AJOUSLI\Lib\site-packages\pybamm\expression_tree\symbol.py", line 673, in __bool__
    raise NotImplementedError("Boolean operator not defined for Symbols.")
NotImplementedError: Boolean operator not defined for Symbols.

Do you know how I can do this?

Your answers would be very helpful to me :)
Thanks for your consideration.

**Comment by valentinsulzer** [comment]
There's two things happening here:
1. The "Total heat transfer coefficient [W.m-2.K-1]" does not expect a function, but a constant. It's not possible to pass in a function of height using the built-in pybamm models. You would have to write your own model that allows it. However, you can specify different heat transfer coefficients for different parts of the cell (edge, top, bottom, tab) if you use the option `{"cell geometry": "pouch"}`.
2. Don't use `if`, `elif`, instead use heavisides: `return (y < 1) * 470 + (y>=1) * 520`.

In your case, if 470 and 520 are really the true values, I wouldn't bother with any of this and just use a single value that is the appropriate area-weighted average of the two. The difference between 470 and 520 is much smaller than the error introduced by some of the simplifying assumptions in the model

**Reply by jouslin** [reply]
Thanks you for your answers !
Is it possible to directly change the Total heat transfer coefficient properties in the folder, similar to how you explained changing the current collector conductivity? (see below)

"This is because the current collector conductivity is a Parameter, not a FunctionParameter
It needs to be change from [this](https://github.com/pybamm-team/PyBaMM/blob/6b642a6c3c9d5d90c15f9a807f8bef5998efda6b/pybamm/parameters/lithium_ion_parameters.py#L282) to [this](https://github.com/pybamm-team/PyBaMM/blob/6b642a6c3c9d5d90c15f9a807f8bef5998efda6b/pybamm/parameters/lithium_ion_parameters.py#L345)"

**Reply by valentinsulzer** [reply]
Yes, if you want to open a PR to do that go ahead. However, this will only make a difference if you're using the option "dimensionality" of 1 or 2 (not the default 0). If dimensionality is 0, the current collector gets lumped, and a change in the y dimension has no effect.

@rtimms has thought more than me about y-z thermal effects and may have further recommendations

**Reply by TomTranter** [reply]
Those heat transfer coefficients are very high. Are you modelling a phase change material? Spatially dependent heat transfer is possible using jellybamm which is something I wrote a few years ago to do distributed thermal models from CT images. Only developed 2d so far but could be used as a starting point for your use case if you are expecting high internal thermal gradients https://jellybamm.readthedocs.io/en/latest/

**Reply by rtimms** [reply, marked as answer]
You can already do this for pouch geometries by setting the parameters “ Edge heat transfer coefficient [W.m-2.K-1]“ (see https://github.com/pybamm-team/PyBaMM/blob/afbf21d45d070e4620cb946fa06c996d60a14f86/pybamm/parameters/thermal_parameters.py#L60) and “ f"{Domain} current collector surface heat transfer coefficient [W.m-2.K-1]” (see https://github.com/pybamm-team/PyBaMM/blob/afbf21d45d070e4620cb946fa06c996d60a14f86/pybamm/parameters/thermal_parameters.py#L104). These both depend on y and z. 

The total heat transfer is only used in lumped thermal models, so cannot depend on space. See  https://github.com/pybamm-team/PyBaMM/blob/develop/docs/source/examples/notebooks/models/thermal-models.ipynb for an overview of the thermal models in pybamm.

**Reply by jouslin** [reply]
Thank you for all these answers, now it's working :)
